File: frontend/src/features/kis/parser.js
/**
 * Frontend deterministic KIS composer command parser.
 *
 * Provides immediate structural preview and validation feedback mirroring the backend grammar.
 * This parser produces preview classifications but does not replace authoritative server-side resolution.
 */

const MALFORMED_PREFIX_REGEX = /^E(?=[0-9+\-: \t])[^\r\n:]*:/m;

/**
 * Extract event patches from an explicit E#: block.
 *
 * @param {string} text - Composer input text
 * @returns {Array<{ event_id: string, instruction: string }>}
 */
export const extractEventPatches = (text) => {
  const normalized = (text || '').trim();
  const headerRegex = /^E([1-9]\d*):[ \t]*(.*)$/gm;
  const headers = [];
  let match;
  while ((match = headerRegex.exec(normalized)) !== null) {
    headers.push({
      index: match.index,
      fullMatch: match[0],
      eventId: `E${match[1]}`,
      afterColon: match[2],
    });
  }

  const patches = [];
  for (let i = 0; i < headers.length; i += 1) {
    const h = headers[i];
    const nextStart = i + 1 < headers.length ? headers[i + 1].index : normalized.length;
    const instruction = (
      h.afterColon + normalized.slice(h.index + h.fullMatch.length, nextStart)
    ).trim();
    patches.push({ event_id: h.eventId, instruction });
  }
  return patches;
};

/**
 * Extract instruction from an /llm-rewrite command block.
 *
 * @param {string} text - Composer input text
 * @returns {string}
 */
export const extractGlobalRewriteInstruction = (text) => {
  const lines = (text || '').trim().split('\n');
  return lines.slice(1).join('\n').trim();
};

/**
 * Parse and validate a KIS composer draft against base intent.
 *
 * @param {string} draft - Raw composer text
 * @param {object|null} baseIntent - Current intent or null
 * @returns {{
 *   kind: 'initial_resolve' | 'patch_events' | 'global_rewrite' | 'invalid',
 *   affectedEventIds: string[],
 *   error: string | null,
 * }}
 */
export const parseComposerDraft = (draft, baseIntent = null) => {
  const normalized = (draft || '').trim();
  if (!normalized) {
    return {
      kind: 'invalid',
      affectedEventIds: [],
      error: 'KIS command must contain non-empty text',
    };
  }

  // Global rewrite command: /llm-rewrite\n<instruction>
  if (normalized.startsWith('/llm-rewrite')) {
    const lines = normalized.split('\n');
    if (lines[0].trim() !== '/llm-rewrite') {
      return {
        kind: 'invalid',
        affectedEventIds: [],
        error: 'Malformed /llm-rewrite command',
      };
    }
    const instruction = lines.slice(1).join('\n').trim();
    if (!instruction) {
      return {
        kind: 'invalid',
        affectedEventIds: [],
        error: '/llm-rewrite requires a non-empty instruction',
      };
    }
    const affectedEventIds = Array.isArray(baseIntent?.events)
      ? baseIntent.events.map((e) => e.id)
      : [];
    return {
      kind: 'global_rewrite',
      affectedEventIds,
      error: null,
    };
  }

  // Header matching for E1..Ek:
  const headerRegex = /^E([1-9]\d*):[ \t]*(.*)$/gm;
  const headers = [];
  let match;
  while ((match = headerRegex.exec(normalized)) !== null) {
    headers.push({
      index: match.index,
      fullMatch: match[0],
      eventNum: parseInt(match[1], 10),
      eventId: `E${match[1]}`,
      afterColon: match[2],
    });
  }

  const baseEventCount = Array.isArray(baseIntent?.events) ? baseIntent.events.length : 0;

  if (headers.length === 0) {
    if (MALFORMED_PREFIX_REGEX.test(normalized)) {
      return {
        kind: 'invalid',
        affectedEventIds: [],
        error: 'Malformed E#: event prefix',
      };
    }
    if (baseEventCount > 0) {
      return {
        kind: 'invalid',
        affectedEventIds: [],
        error: 'Existing KIS intents require explicit E#: event instructions',
      };
    }
    return {
      kind: 'initial_resolve',
      affectedEventIds: [],
      error: null,
    };
  }

  if (headers[0].index !== 0) {
    return {
      kind: 'invalid',
      affectedEventIds: [],
      error: 'Explicit E#: event instructions must begin the command',
    };
  }

  // Check for any malformed event prefixes in the text that did not match clean headers
  const malformedGlobal = /^E(?=[0-9+\-: \t])[^\r\n:]*:/gm;
  const validStarts = new Set(headers.map((h) => h.index));
  let malMatch;
  while ((malMatch = malformedGlobal.exec(normalized)) !== null) {
    if (!validStarts.has(malMatch.index)) {
      return {
        kind: 'invalid',
        affectedEventIds: [],
        error: 'Malformed E#: event prefix',
      };
    }
  }

  // Validate instructions non-empty
  for (let i = 0; i < headers.length; i += 1) {
    const h = headers[i];
    const nextStart = i + 1 < headers.length ? headers[i + 1].index : normalized.length;
    const instruction = (
      h.afterColon + normalized.slice(h.index + h.fullMatch.length, nextStart)
    ).trim();
    if (!instruction) {
      return {
        kind: 'invalid',
        affectedEventIds: [],
        error: `${h.eventId} requires an instruction`,
      };
    }
  }

  // Validate event numbering
  const numbers = headers.map((h) => h.eventNum);
  if (new Set(numbers).size !== numbers.length) {
    return {
      kind: 'invalid',
      affectedEventIds: [],
      error: 'Duplicate event IDs are not allowed',
    };
  }

  for (let i = 1; i < numbers.length; i += 1) {
    if (numbers[i] <= numbers[i - 1]) {
      return {
        kind: 'invalid',
        affectedEventIds: [],
        error: 'Event headers must be in strictly increasing order',
      };
    }
  }

  const newNumbers = numbers.filter((n) => n > baseEventCount);
  if (newNumbers.length > 0) {
    const maxNew = newNumbers[newNumbers.length - 1];
    const expected = [];
    for (let n = baseEventCount + 1; n <= maxNew; n += 1) {
      expected.push(n);
    }
    const newSet = new Set(newNumbers);
    const missing = expected.find((n) => !newSet.has(n));
    if (missing !== undefined) {
      return {
        kind: 'invalid',
        affectedEventIds: [],
        error: `Missing E${missing} before a new event`,
      };
    }
  }

  const affectedEventIds = headers.map((h) => h.eventId);
  if (baseEventCount === 0) {
    return {
      kind: 'initial_resolve',
      affectedEventIds,
      error: null,
    };
  }

  return {
    kind: 'patch_events',
    affectedEventIds,
    error: null,
  };
};

File: src/hcmai/kis/parser.py
"""Parse deterministic KIS composer commands without model inference.

This module recognizes initial natural-language input, ordered event-scoped
instructions, and the explicit global rewrite command. It only validates command
grammar and event numbering; semantic interpretation belongs to KIS resolvers.
"""

from __future__ import annotations

from dataclasses import dataclass
import re
from typing import Literal, TypeAlias


_EVENT_HEADER = re.compile(r"(?m)^E([1-9]\d*):[ \t]*(.*)$")
_MALFORMED_EVENT_PREFIX = re.compile(
    r"(?m)^E(?=[0-9+\-: \t])[^\r\n:]*:"
)


class KISCommandError(ValueError):
    """Raised when composer text does not satisfy the deterministic grammar."""


@dataclass(frozen=True, slots=True)
class EventPatchInstruction:
    """One user-authored semantic instruction scoped to a canonical event ID."""

    event_id: str
    instruction: str


@dataclass(frozen=True, slots=True)
class ParsedInitialNatural:
    """An initial unscoped natural-language request."""

    kind: Literal["initial_natural"] = "initial_natural"
    text: str = ""


@dataclass(frozen=True, slots=True)
class ParsedInitialExplicit:
    """An initial contiguous batch whose event IDs are explicitly supplied."""

    kind: Literal["initial_explicit"] = "initial_explicit"
    patches: tuple[EventPatchInstruction, ...] = ()


@dataclass(frozen=True, slots=True)
class ParsedPatchEvents:
    """An ordered batch of semantic updates to named events."""

    kind: Literal["patch_events"] = "patch_events"
    patches: tuple[EventPatchInstruction, ...] = ()


@dataclass(frozen=True, slots=True)
class ParsedGlobalRewrite:
    """An explicit instruction authorizing global semantic rewriting."""

    kind: Literal["global_rewrite"] = "global_rewrite"
    instruction: str = ""


ParsedKISCommand: TypeAlias = (
    ParsedInitialNatural
    | ParsedInitialExplicit
    | ParsedPatchEvents
    | ParsedGlobalRewrite
)


def parse_kis_command(text: str, *, base_event_count: int) -> ParsedKISCommand:
    """Classify and validate one KIS composer submission deterministically.

    Existing intents require an explicit ``E#:`` target for local changes. New
    event IDs must extend the existing timeline contiguously, while all supplied
    headers must already be in strictly increasing user order.
    """
    if base_event_count < 0:
        raise ValueError("base_event_count must be non-negative")

    normalized = text.strip()
    if not normalized:
        raise KISCommandError("KIS command must contain non-empty text")

    if normalized.startswith("/llm-rewrite"):
        return _parse_global_rewrite(normalized)

    headers = tuple(_EVENT_HEADER.finditer(normalized))
    if not headers:
        if _MALFORMED_EVENT_PREFIX.search(normalized):
            raise KISCommandError("Malformed E#: event prefix")
        if base_event_count:
            raise KISCommandError(
                "Existing KIS intents require explicit E#: event instructions"
            )
        return ParsedInitialNatural(text=normalized)

    if headers[0].start() != 0:
        raise KISCommandError("Explicit E#: event instructions must begin the command")
    _reject_malformed_event_prefixes(normalized, headers)
    patches = _parse_event_patches(normalized, headers)
    _validate_event_ids(patches, base_event_count=base_event_count)

    if base_event_count == 0:
        return ParsedInitialExplicit(patches=patches)
    return ParsedPatchEvents(patches=patches)


def _parse_global_rewrite(text: str) -> ParsedGlobalRewrite:
    """Validate the dedicated two-line global rewrite command."""
    lines = text.splitlines()
    if lines[0] != "/llm-rewrite":
        raise KISCommandError("Malformed /llm-rewrite command")
    instruction = "\n".join(lines[1:]).strip()
    if not instruction:
        raise KISCommandError("/llm-rewrite requires a non-empty instruction")
    return ParsedGlobalRewrite(instruction=instruction)


def _reject_malformed_event_prefixes(
    text: str, headers: tuple[re.Match[str], ...]
) -> None:
    """Reject event-like lines that the strict header expression did not accept."""
    valid_starts = {header.start() for header in headers}
    for malformed in _MALFORMED_EVENT_PREFIX.finditer(text):
        if malformed.start() not in valid_starts:
            raise KISCommandError("Malformed E#: event prefix")


def _parse_event_patches(
    text: str, headers: tuple[re.Match[str], ...]
) -> tuple[EventPatchInstruction, ...]:
    """Collect each header's content through the start of its next header."""
    patches: list[EventPatchInstruction] = []
    for index, header in enumerate(headers):
        next_start = headers[index + 1].start() if index + 1 < len(headers) else len(text)
        instruction = (header.group(2) + text[header.end() : next_start]).strip()
        if not instruction:
            raise KISCommandError(f"{header.group(0).split(':', 1)[0]} requires an instruction")
        patches.append(
            EventPatchInstruction(event_id=f"E{header.group(1)}", instruction=instruction)
        )
    return tuple(patches)


def _validate_event_ids(
    patches: tuple[EventPatchInstruction, ...], *, base_event_count: int
) -> None:
    """Enforce duplicate-free, ordered headers and contiguous newly added IDs."""
    numbers = [int(patch.event_id[1:]) for patch in patches]
    if len(numbers) != len(set(numbers)):
        raise KISCommandError("duplicate event IDs are not allowed")
    if numbers != sorted(numbers):
        raise KISCommandError("event headers must be in strictly increasing order")

    new_numbers = [number for number in numbers if number > base_event_count]
    if new_numbers:
        expected = list(range(base_event_count + 1, new_numbers[-1] + 1))
        if new_numbers != expected:
            missing = next(number for number in expected if number not in new_numbers)
            raise KISCommandError(f"missing E{missing} before a new event")


__all__ = [
    "EventPatchInstruction",
    "KISCommandError",
    "ParsedGlobalRewrite",
    "ParsedInitialExplicit",
    "ParsedInitialNatural",
    "ParsedKISCommand",
    "ParsedPatchEvents",
    "parse_kis_command",
]
